### PR TITLE
Extend cilium-operator binary to be used as command line tool

### LINKED ISSUE #13132: Expose cilium-operator metrics
We can view metrics from the Cilium CLI via `cilium metrics list`, however the same doesn't exist for cilium-operator. Internally, the metrics are being tracked; just not exposed.

### PR DESCRIPTION
**Pull request by fristonio**
This PullRequest aims at extending the `cilium-operator` binary to be used as a command-line tool. There are two reworks associated with this PR.
1. Make `cilium-operator` to use `openapi` config as apiserver implementation similar to what we have for Cilium and `cilium-health`.
2. Add metric subcommand to `cilium-operator` to list metrics from a running cilium-operator instance apiserver.

Fixes #13132

image

### CODE REVIEW

**Comment by fristonio** [comment]
test-me-please

**Comment by fristonio** [comment]
retest-gke

File: api/v1/operator/openapi.yaml
Comment (sayboras) [inline api/v1/operator/openapi.yaml]: Are we following the convention in which basepath is same as version? This https://github.com/cilium/cilium/blob/3b56fc99a42ad6961fdd30a69e211e5777b36d9f/api/v1/openapi.yaml#L9 seems to have user friendly `v1` for basepath only.

File: api/v1/operator/openapi.yaml
Comment (sayboras) [inline api/v1/operator/openapi.yaml]: nit: new line :thinking:

File: operator/main.go
Comment (sayboras) [inline operator/main.go:117]: I find it's a little bit confusing to have same binary for CLI + deamon server, it's sometimes confusing to me. Keen to hear inputs from others.

File: operator/cmd/root.go
Comment (sayboras) [inline operator/cmd/root.go:32]: This one is interesting. Compared to cilium-agent, operator only runs in one or two nodes (in HA mode), so are we expecting end user to run `operator metrics list` in one of these nodes ? Is there any difference between active and standby node ? if yes, user might need to know the leader node to run this command.

**Review by sayboras — COMMENTED** [review]
TIL about go-swagger generation :+1: 

Couple of comments I have as below, will do another round of review later.

File: api/v1/operator/openapi.yaml
Comment (fristonio) [reply api/v1/operator/openapi.yaml]: Since its an initial version for the operator API I wanted to keep it as `v1beta` similar to what we have for `cilium-health`. I think we can change the base path to `/v1` if everyone agrees.

File: operator/cmd/root.go
Comment (fristonio) [reply operator/cmd/root.go:32]: So the metrics will be listed individually for the two instances in HA mode. For the leader operator we can have extra metrics registered like `identity_gc_*` along with the standard process metrics. The idea was to provide some sort of method within `cilium-operator` container to gather information about the operator, initially metrics.
I think what you said makes sense and we can also add a status command for the operator which can provide status information of the underlying instance. :rocket: 
I can take that up as a follow-up issue.

File: operator/cmd/metrics_list.go
Comment (aanm) [inline operator/cmd/metrics_list.go:15]: Missing copyright header

File: operator/cmd/root.go
Comment (aanm) [inline operator/cmd/root.go:15]: Missing copyright header

File: operator/cmd/root.go
Comment (sayboras) [reply operator/cmd/root.go:32]: thanks, sounds good to me :+1:

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
Comment (aanm) [inline install/kubernetes/cilium/templates/cilium-operator-deployment.yaml]: Can we avoid changing this path?

**Review by aanm — CHANGES_REQUESTED** [review]

File: api/v1/operator/client/cilium_operator_client.go
Comment (christarazi) [inline api/v1/operator/client/cilium_operator_client.go:1]: (Noting the commit title here, not the actual change)

Commit title says "operator apiserver" which is slightly confusing to me as "apiserver" usually refers to the K8s apiserver. Could you clarify the title or simply avoid using "apiserver"? Maybe you meant "the API served by the Operator"?

File: operator/api/server.go
Comment (christarazi) [inline operator/api/server.go]: This should refer to the Operator, rather than the agent?

**Review by christarazi — CHANGES_REQUESTED** [review]
Looks good! A few more minor comments on top of everyone else's review

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
Comment (fristonio) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml]: We can set the base path for open API spec to `/` and then we can avoid this change. What do you think should  be the preferred basePath for the cilium-operator API?

File: operator/api/server.go
Comment (fristonio) [reply operator/api/server.go]: Ah, this function was taken from the previous implementation that lived inside the operator. I can update the comment. :sweat_smile:

File: api/v1/operator/client/cilium_operator_client.go
Comment (fristonio) [reply api/v1/operator/client/cilium_operator_client.go:1]: Done

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
Comment (aanm) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml]: Can't we have an alias path to also serve `/healthz`?

Comment (fristonio) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml]: One way I see is to initialize the `http.Server` using the default mux with a specific entry for `/healthz` and the default handler will fallback to open-api router.
I will check if this works.

**Review by christarazi — APPROVED** [review]
LGTM besides already pending items

File: api/v1/operator/openapi.yaml
Comment (fristonio) [reply api/v1/operator/openapi.yaml]: Discussed offline with @aanm 
It makes sense to keep the basepath as `/v1` but the version should be `v1beta`.

File: install/kubernetes/cilium/templates/cilium-operator-deployment.yaml
Comment (fristonio) [reply install/kubernetes/cilium/templates/cilium-operator-deployment.yaml]: Seems to work

image

**Review by kaworu — APPROVED** [review]
Overall LGTM. I think `CODEOWNERS` should be updated to take into account the new `/api/v1/operator` directory:

```diff
diff --git a/CODEOWNERS b/CODEOWNERS
index bdbe06c83..29d6a492e 100644
--- a/CODEOWNERS
+++ b/CODEOWNERS
@@ -35,6 +35,7 @@
 /api/ @cilium/api
 /api/v1/flow/ @cilium/api @cilium/hubble
 /api/v1/observer/ @cilium/api @cilium/hubble
+/api/v1/operator/ @cilium/api @cilium/operator
 /api/v1/peer/ @cilium/api @cilium/hubble
 /bpf/ @cilium/bpf
 Makefile* @cilium/build
```

**Comment by fristonio** [comment]
test-me-please

File: api/v1/operator/openapi.yaml
Comment (aanm) [inline api/v1/operator/openapi.yaml]: This should not have a schema.

File: operator/api/server.go
Comment (aanm) [inline operator/api/server.go]: I was almost missing this, setting the header in here made me double check. The current master is simply returning `ok` with content-type `plain/text`. This healthz should not be modified in the `/healthz` nor `/v1/healthz`.

```
$ curl -vvv 127.0.0.1:9234/healthz
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 9234 (#0)
> GET /healthz HTTP/1.1
> Host: 127.0.0.1:9234
> User-Agent: curl/7.58.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Date: Tue, 05 Jan 2021 16:49:13 GMT
< Content-Length: 2
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host 127.0.0.1 left intact
ok
```

**Review by aanm — CHANGES_REQUESTED** [review]

File: api/v1/operator/openapi.yaml
Comment (fristonio) [reply api/v1/operator/openapi.yaml]: Ok, I did not realize that we need to preserve the previously used plain text schema for `/healthz`

File: operator/api/server.go
Comment (fristonio) [reply operator/api/server.go]: I have pushed the required changes to keep the schema the same as before. 
image

**Comment by fristonio** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Review by sayboras — APPROVED** [review]
LGTM 💯

**Pull request merged**
